Install failure disables run button

Please ensure you have given all the following requested information in your report.

#### Issue details
If running the application fails, the Run button is disabled, requiring the project to be closed and opened again to re-enable it.
#### Reproduction steps/code
Create a robovm run configuration targeting a hardware device. Unplug any devices from the computer and click the Run button. The run will fail with "[ERROR] Couldn't start application
java.lang.RuntimeException: No devices connected"

Now the Run button is disabled (greyed out.) Closing the project and reopening it will re-enable the button.

#### Configuration
Robovm 2.3.4-SNAPSHOT, AS 3.0.

**Build Tools:**
- [x ] IDEA plugin
- [ ] Eclipse plugin
- [ ] Gradle plugin
 
**Versions:**

_Please provide the version of RoboVM, XCode and JDK used_
- Robovm: 2.3.4-SNAPSHOT
- XCode: 9.2
- JDK: 1.8

**Build Targets:**

Hardware iOS devices, 32 or 64 bit.
---

--- a/compiler/compiler/src/main/java/org/robovm/compiler/target/ios/IOSTarget.java
+++ b/compiler/compiler/src/main/java/org/robovm/compiler/target/ios/IOSTarget.java
@@ -357,8 +357,8 @@ public class IOSTarget extends AbstractTarget {
                 // Copy the provisioning profile
                 copyProvisioningProfile(provisioningProfile, installDir);
                 boolean getTaskAllow = provisioningProfile.getType() == Type.Development;
-                signFrameworks(installDir, getTaskAllow);
-                signAppExtensions(installDir, getTaskAllow);
+                signFrameworks(signIdentity, installDir);
+                signAppExtensions(signIdentity, installDir);
                 codesignApp(signIdentity, getOrCreateEntitlementsPList(getTaskAllow, getBundleId()), installDir);
             }
         }
@@ -391,35 +391,41 @@ public class IOSTarget extends AbstractTarget {
             } else {
                 copyProvisioningProfile(provisioningProfile, appDir);
                 boolean getTaskAllow = provisioningProfile.getType() == Type.Development;
-                signFrameworks(appDir, getTaskAllow);
-                signAppExtensions(appDir, getTaskAllow);
+                signFrameworks(signIdentity, appDir);
+                signAppExtensions(signIdentity, appDir);
                 // sign the app
                 codesignApp(signIdentity, getOrCreateEntitlementsPList(getTaskAllow, getBundleId()), appDir);
+            }
+        } else { // simulator
+            if (sdk.getVersionCode() >= 0x0B0300) {
+                // code signing of frameworks and app extensions are required since iOS 11.3
+                signFrameworks(SigningIdentity.ADHOC, appDir);
+                signAppExtensions(SigningIdentity.ADHOC, appDir);
             }
         }
     }
 
-    private void signFrameworks(File appDir, boolean getTaskAllow) throws IOException {
+    private void signFrameworks(SigningIdentity identity, File appDir) throws IOException {
         // sign dynamic frameworks first
         File frameworksDir = new File(appDir, "Frameworks");
         if (frameworksDir.exists() && frameworksDir.isDirectory()) {
             // Sign swift rt libs
             for (File swiftLib : frameworksDir.listFiles()) {
                 if (swiftLib.getName().endsWith(".dylib")) {
-                    codesignSwiftLib(signIdentity, swiftLib);
+                    codesignSwiftLib(identity, swiftLib);
                 }
             }
 
             // sign embedded frameworks
             for (File framework : frameworksDir.listFiles()) {
                 if (framework.isDirectory() && framework.getName().endsWith(".framework")) {
-                    codesignCustomFramework(signIdentity, framework);
+                    codesignCustomFramework(identity, framework);
                 }
             }
         }
     }
 
-    private void signAppExtensions(File appDir, boolean getTaskAllow) throws IOException {
+    private void signAppExtensions(SigningIdentity identity, File appDir) throws IOException {
         // sign dynamic frameworks first
         File extensionsDir = new File(appDir, "PlugIns");
         if (extensionsDir.exists() && extensionsDir.isDirectory()) {
@@ -427,7 +433,7 @@ public class IOSTarget extends AbstractTarget {
             for (File extension : extensionsDir.listFiles()) {
                 if (extension.isDirectory() && extension.getName().endsWith(".appex")) {
                     // now sign
-                    codesignAppExtension(signIdentity, extension);
+                    codesignAppExtension(identity, extension);
                 }
             }
         }

--- a/compiler/compiler/src/main/java/org/robovm/compiler/target/ios/SigningIdentity.java
+++ b/compiler/compiler/src/main/java/org/robovm/compiler/target/ios/SigningIdentity.java
@@ -33,7 +33,8 @@ public class SigningIdentity implements Comparable<SigningIdentity> {
 
     private final String name;
     private final String fingerprint;
-    
+    public static final SigningIdentity ADHOC = new SigningIdentity("ad-hoc (simulator only)", "-");
+
     SigningIdentity(String name, String fingerprint) {
         this.name = name;
         this.fingerprint = fingerprint;

--- a/plugins/idea/src/main/java/org/robovm/idea/RoboVmPlugin.java
+++ b/plugins/idea/src/main/java/org/robovm/idea/RoboVmPlugin.java
@@ -45,6 +45,7 @@ import com.intellij.ui.content.Content;
 import com.intellij.util.ui.UIUtil;
 import org.apache.commons.compress.archivers.ArchiveEntry;
 import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
+import org.apache.commons.io.FileUtils;
 import org.apache.commons.io.IOUtils;
 import org.jetbrains.annotations.NotNull;
 import org.robovm.compiler.Version;
@@ -416,8 +417,9 @@ public class RoboVmPlugin {
     private static void extractArchive(String archive, File dest) {
         archive = "/" + archive;
         TarArchiveInputStream in = null;
-        boolean isSnapshot = Version.getVersion().toLowerCase().contains("snapshot");
+//        boolean isSnapshot = Version.getVersion().toLowerCase().contains("snapshot");
         try {
+            boolean filesWereUpdated = false;
             in = new TarArchiveInputStream(new GZIPInputStream(RoboVmPlugin.class.getResourceAsStream(archive)));
             ArchiveEntry entry = null;
             while ((entry = in.getNextEntry()) != null) {
@@ -425,7 +427,8 @@ public class RoboVmPlugin {
                 if (entry.isDirectory()) {
                     f.mkdirs();
                 } else {
-                    if(!isSnapshot && f.exists()) {
+                    // skip extracting if file looks to be same as it archive (ts and size matches)
+                    if (f.exists() && f.lastModified() == entry.getLastModifiedDate().getTime() && f.length() == entry.getSize()) {
                         continue;
                     }
                     f.getParentFile().mkdirs();
@@ -442,11 +445,22 @@ public class RoboVmPlugin {
                         if (out != null)
                             IOUtils.closeQuietly(out);
                     }
+
+                    // mark that there was a change to SDK files
+                    filesWereUpdated = true;
                 }
             }
             logInfo(null, "Installed RoboVM SDK %s to %s", Version.getVersion(), dest.getAbsolutePath());
 
-            // make all files in bin executable
+            if (filesWereUpdated) {
+                File cacheLog = new File(System.getProperty("user.home"), ".robovm/cache");
+                logInfo(null, "Clearing cache log folder due SDK files changed.");
+                try {
+                    FileUtils.deleteDirectory(cacheLog);
+                } catch (IOException ignored) {
+                }
+            }
+
             for (File file : new File(getSdkHome(), "bin").listFiles()) {
                 file.setExecutable(true);
             }

--- a/plugins/idea/src/main/java/org/robovm/idea/running/RoboVmRunProfileState.java
+++ b/plugins/idea/src/main/java/org/robovm/idea/running/RoboVmRunProfileState.java
@@ -127,11 +127,11 @@ public class RoboVmRunProfileState extends CommandLineState {
             } else if (getEnvironment().getExecutor().getId().equals(RoboVmRunner.DEBUG_EXECUTOR)) {
                 return executeRun();
             } else {
-                return null;
+                throw new ExecutionException("Unsupported executor " + getEnvironment().getExecutor().getId());
             }
         } catch(Throwable t) {
             RoboVmPlugin.logErrorThrowable(getEnvironment().getProject(), "Couldn't start application", t, true);
-            return null;
+            throw new ExecutionException(t);
         }
     }
 

